scipy/special/ellint_carlson_cpp_lite/_rj.hh

```diff
@@ -0,0 +1,701 @@
+
+
+/* The notation follows the one used in the DLMF:
+ * https://dlmf.nist.gov/19.27.i
+ */
+template<typename T>
+struct AsymConfig
+{
+    T a;
+    T b;
+    T c;
+    T f;
+    T g;
+    T h;
+};
+
+template<typename T>
+static typing::real_only<T, AsymFlag>
+rj_asym_conf(const T& x, const T& y, const T& z, const T& p,
+             AsymConfig<T>& conf)
+{
+    T t;
+
+    t = z / p;
+    if ( asymp_zero(t) )
+    {
+	conf.c = (x + y + z) / 3.0;
+	return AsymFlag::hugep;
+    }
+
+    /* this bound is sharp. RJ in this case behaves with logarithmic
+     * singularity as p -> +0 */
+    if ( abs_close_zero(p) || ( (x != 0.0) && asymp_zero(p / x) ) )
+    {
+	conf.f = std::sqrt(x * y * z);
+	return AsymFlag::tinyp;
+    }
+
+    /* XXX until RJ(0, y, z, p) is implemented, this is useless */
+    /*
+    t = (*x) / fmin((*y), (*p));
+    if ( ( 0.0 < (*x) && (*x) <= 1e-26 ) || ASYMP_ZERO(t) )
+    {
+	*h = sqrt((*y) * (*z));
+	if ( (*h) / (*p) + 0.5 * ((*y) + (*z)) / (*h) <= sqrt((*h) / (*x)) )
+	{
+	    return asymp_tinyx;
+	}
+    }
+    */
+
+    t = y / std::fmin(z, p);
+    if ( ( (0.0 < y) && (y <= 1e-26) ) || asymp_zero(t) )
+    {
+	/* bound fairly sharp even if p is large */
+	conf.a = (T)0.5 * (x + y);
+	conf.g = std::sqrt(x * y);
+	if ( (conf.a / z + conf.a / p) * std::abs(std::log(p / conf.a)) <=
+	     1.0 )
+	{
+	    return AsymFlag::tinyy;
+	}
+    }
+
+    if ( (x != 0.0) && asymp_zero(std::fmax(z, p) / x) )
+    {
+	/* bound might not be sharp if x + 2p much larger than (yz) ** 2,
+	 * but this is unlikely to be true anyway. */
+	return AsymFlag::hugey;
+    }
+
+    if ( (z != 0.0) && asymp_zero(std::fmax(y, p) / z) )
+    {
+	conf.b = 0.5 * (x + y);
+	conf.h = std::sqrt(x * y);
+	/* when bounds are sharp */
+	if ( std::abs(std::log(z / (conf.b + conf.h))) <= std::sqrt(z) )
+	{
+	    return AsymFlag::hugez;
+	}
+    }
+
+    return AsymFlag::nothing;
+}
+
+
+/* Prevent division by zero due to underflow in atan(sqrt(z)) / sqrt(z) and
+ * square root of negative number in the real context. */
+template<typename T>
+static inline typing::cplx_only<T, T>
+safe_atan_sqrt_div(T z)
+{
+    if ( argcheck::too_small(z) )
+    {
+	return T(1.0);
+    }
+    T s = std::sqrt(z);
+    return std::atan(s) / s;
+}
+
+template<typename T>
+static inline typing::real_only<T, T>
+safe_atan_sqrt_div(T x)
+{
+    if ( argcheck::too_small(x) )
+    {
+	return T(1.0);
+    }
+    T s;
+    if ( x < 0.0 )
+    {
+	s = std::sqrt(-x);
+	return std::atanh(s) / s;
+    }
+    s = std::sqrt(x);
+    return std::atan(s) / s;
+}
+
+
+}  /* namespace ellint_carlson::rjimpl */
+
+template<typename T>
+ExitStatus
+rj(const T& x, const T& y, const T& z, const T& p, const double& rerr, T& res,
+   bool noasymp)
+{
+    typedef typing::decplx_t<T> RT;
+
+    rjimpl::ArgCases classify;
+    T cct1[6];
+    T cct2[6];
+
+    ExitStatus status = ExitStatus::success;
+#ifndef ELLINT_NO_VALIDATE_RELATIVE_ERROR_BOUND
+    if ( argcheck::invalid_rerr(rerr, 1.0e-4) )
```

Comment:
There are 4 different usages of `argcheck::invalid_rerr` each using its own presumably tuned value -- is there an indication of how these values were chosen?